krun/platform.py

```diff
@@ -0,0 +1,129 @@
+# Various CPU related abstrations
+
+import time
+import os
+import sys
+from collections import OrderedDict
+from krun.util import fatal, collect_cmd_output
+
+
+class BasePlatform(object):
+    CPU_TEMP_MANDATORY_WAIT = 1
+    CPU_TEMP_POLL_FREQ = 5
+    CPU_TEMP_POLLS_BEFORE_MELTDOWN = 12  #  * 5 = one minute
+
+    def __init__(self):
+        self.audit = OrderedDict()
+
+    def wait_until_cpu_cool(self):
+        time.sleep(BasePlatform.CPU_TEMP_MANDATORY_WAIT)
+        msg_shown = False
+        trys = 0
+        while True:
+            cool, reason = self.has_cpu_cooled()
+            if cool:
+                break
+
+            # if we get here, CPU is too hot!
+            if not msg_shown:
+                print("CPU is running hot.")
+                print(reason)
+                sys.stdout.write("Waiting to cool")
+                sys.stdout.flush()
+                msg_shown = True
+
+            trys += 1
+            if trys >= BasePlatform.CPU_TEMP_POLLS_BEFORE_MELTDOWN:
+                print("")
+                fatal("CPU didn't cool down")
+
+            time.sleep(BasePlatform.CPU_TEMP_POLL_FREQ)
+            sys.stdout.write(".")
+            sys.stdout.flush()
+        print("")
+
+    # When porting to a new platform, implement the following:
+    def take_cpu_temp_readings(self):
+        raise NotImplementedError("abstract")
+
+    def set_base_cpu_temps(self):
+        raise NotImplementedError("abstract")
+
+    def has_cpu_cooled(self):
+        raise NotImplementedError("abstract")
+
+    def check_cpu_throttled(self):
+        raise NotImplementedError("abstract")
+
+    # And you may want to extend this
+    def collect_audit(self):
+        self.audit["uname"] = collect_cmd_output("uname")
+        self.audit["dmesg"] = collect_cmd_output("dmesg")
+
+class LinuxPlatform(BasePlatform):
+    THERMAL_BASE = "/sys/class/thermal/"
+    CPU_GOV_FILE = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
+
+    # Temperature files under /sys measure in millidegrees
+    # https://www.kernel.org/doc/Documentation/thermal/sysfs-api.txt
+    THRESHOLD = 1000  # therefore one degree
+
+    def __init__(self):
+        self.initial_readings = None
+        self.zones = self._find_thermal_zones()
+        BasePlatform.__init__(self)
+
+    def _find_thermal_zones(self):
+        return [x for x in os.listdir(LinuxPlatform.THERMAL_BASE) if
+                x.startswith("thermal_zone")]
+
+    def set_base_cpu_temps(self):
+        self.initial_readings = self.take_cpu_temp_readings()
+
+    def _read_zone(self, zone):
+        fn = os.path.join(LinuxPlatform.THERMAL_BASE, zone, "temp")
+        with open(fn, "r") as fh:
+            return int(fh.read())
+
+    def take_cpu_temp_readings(self):
+        return [self._read_zone(z) for z in self.zones]
+
+    def has_cpu_cooled(self):
+        """returns tuple: cool * str_reason_if_false"""
```

Comment:
What is `cool' then? I think this isn't a very helpful docstring as it stands. At least, I still don't get it.